frontends/src/terminal/screen_buf.rs

```diff
@@ -0,0 +1,227 @@
+use pretty::{Pos, Rect, Region};
+use pretty::{Shade, Style};
+
+use super::Error;
+
+/// Represents a screen full of characters. It buffers changes to the
+/// characters, and can produce a set of instructions for efficiently updating
+/// the screen to reflect those changes.
+#[derive(Debug)]
+pub struct ScreenBuf {
+    cells: Vec<Vec<CharCell>>,
+}
+
+/// Represents a single character on a screen, with style properties.
+/// It keeps track of whether it has changed since the last time the ScreenBuf was redisplayed.
+#[derive(Clone, Copy, Debug)]
+struct CharCell {
+    ch: char,
+    shade: Shade,
+    style: Style, // TODO except background is ignored?
+    needs_redisplay: bool,
+}
+
+/// Instructions for how to update a terminal.
+#[derive(Clone, Copy)]
+pub enum ScreenOp {
+    Print(char),
+    Apply(Style),
+    Goto(Pos),
+}
+
+/// An iterator that produces instructions for updating a screen to match changes in a ScreenBuf.
+pub struct ScreenBufIter<'a> {
+    buf: &'a mut ScreenBuf,
+    current_style: Option<Style>,
+    cursor_pos: Option<Pos>,
+}
+
+impl ScreenBuf {
+    pub fn new() -> Self {
+        ScreenBuf { cells: Vec::new() }
+    }
+
+    /// Get `ScreenOp` instructions that describe all changes to the screen buffer since the last time this method was called.
+    pub fn drain_changes(&mut self) -> ScreenBufIter {
+        ScreenBufIter {
+            buf: self,
+            current_style: None,
+            cursor_pos: None,
+        }
+    }
+
+    pub fn resize(&mut self, size: Pos) {
+        self.cells = Vec::new();
+        let mut row = Vec::new();
+        row.resize_with(size.col as usize, Default::default);
+        self.cells.resize(size.row as usize, row);
+    }
+
+    pub fn clear(&mut self) {
+        let size = self.size();
+        self.resize(size);
+    }
+
+    pub fn write_str(&mut self, pos: Pos, s: &str, style: Style) -> Result<(), Error> {
+        let mut char_pos = pos;
+        for ch in s.chars() {
+            self.set_char_with_style(char_pos, ch, style)?;
+            char_pos = self.next_pos(char_pos).ok_or(Error::OutOfBounds)?;
+        }
+        Ok(())
+    }
+
+    pub fn shade_region(&mut self, region: Region, shade: Shade) -> Result<(), Error> {
+        let mut shade_rect = |rect: Rect| {
+            for r in rect.rows.0..rect.rows.1 {
+                for c in rect.cols.0..rect.cols.1 {
+                    let pos = Pos { row: r, col: c };
+                    self.set_shade(pos, shade)?;
+                }
+            }
+            Ok(())
+        };
+        shade_rect(region.body())?;
+        shade_rect(region.last_line())
```

Comment:
TODO: probably just inline these two calls (eliminates the closure, and eliminates one for loop in one case), and don't make `struct Rect` public.